File: mastery/react-flow/product/packages/contract/media-span.ts
/**
 * The "Media-Native" Trace Unit.
 * This is the shared contract between the Engine and the Shell.
 */
export interface MediaSpan {
  traceId: string;
  spanId: string;
  timestamp: number; // UTC Epoch ms

  // Media-Specific Metadata
  media: {
    frameType: 'I' | 'P' | 'B';
    gopIndex: number;
    pts: number; // Presentation Timestamp
    dts: number; // Decoding Timestamp
    bitrate: number; // bits per second
    codec: 'h264' | 'h265';
  };

  // Health Invariants
  health: {
    hasJitter: boolean;
    isCorrupt: boolean;
    isDropped: boolean;
    latencyMs: number; // E2E Latency
  };

  // Standard Infrastructure Context
  resource: {
    podId: string;
    nodeId: string;
    cameraId: string;
    streamUrl: string;
  };
}

export type StreamStatus = 'idle' | 'streaming' | 'error' | 'reconnecting';

export interface StreamNode {
  id: string;
  type: 'camera' | 'transcoder' | 'egress';
  status: StreamStatus;
  metrics: {
    fps: number;
    bandwidth: number;
  };
}

File: mastery/react-flow/product/packages/contract/media_span.go
package contract

// MediaSpan represents the "Media-Native" Trace Unit.
type MediaSpan struct {
	TraceID   string `json:"traceId"`
	SpanID    string `json:"spanId"`
	Timestamp int64  `json:"timestamp"`

	Media MediaMetadata `json:"media"`
	Health HealthMetrics `json:"health"`
	Resource ResourceContext `json:"resource"`
}

type MediaMetadata struct {
	FrameType string `json:"frameType"` // "I", "P", "B"
	GOPIndex  int    `json:"gopIndex"`
	PTS       int64  `json:"pts"`
	DTS       int64  `json:"dts"`
	Bitrate   int    `json:"bitrate"`
	Codec     string `json:"codec"`
}

type HealthMetrics struct {
	HasJitter  bool    `json:"hasJitter"`
	IsCorrupt  bool    `json:"isCorrupt"`
	IsDropped  bool    `json:"isDropped"`
	LatencyMs  float64 `json:"latencyMs"`
}

type ResourceContext struct {
	PodID     string `json:"podId"`
	NodeID    string `json:"nodeId"`
	CameraID  string `json:"cameraId"`
	StreamURL string `json:"streamUrl"`
}
